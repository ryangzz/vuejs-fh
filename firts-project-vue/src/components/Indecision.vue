<template>
  <img :src="image" alt="bg" v-if="image">
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" name="pregunta" id="pregunta" placeholder="Hazme una pregunta" v-model="question">
    <p>Recuerda terminar un signo de interrogacion (?)</p>
    <div v-if="quiestionValid">
      <h2>{{question}}</h2>
      <h1>{{answer}}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name:'Indecision',
  data() {
    return {
      question:'',
      answer:'',
      image:null,
      quiestionValid:false
    }
  },
  methods:{
    async getAnswer(){
      try {
        this.answer = "Pensado ...";
      const { answer, image }  = await fetch("https://yesno.wtf/api").then(resp=> resp.json());
      this.answer = answer, this.image = image;
      } catch (error) {
        this.answer = "No se pudo cargar Api";
        this.image = null;
      }
    }
  },
  watch:{
    question(newValue){
      this.quiestionValid = false;
      console.log(newValue);
      if (!newValue.includes("?")) return
      this.getAnswer();
      this.quiestionValid = true
    }
  },
}
</script>

<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>