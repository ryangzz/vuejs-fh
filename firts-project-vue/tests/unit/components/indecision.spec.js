import { shallowMount } from "@vue/test-utils";
import Indecision from '@/components/Indecision.vue';
import { Promise } from "core-js";

describe("Indecision Component", ()=>{
  let wrapper;
  let spyLog;
  global.fetch  = jest.fn(()=> Promise.resolve({
    json:()=>Promise.resolve({
      answer:"Si!",
      forced:false,
      image:"https://yesno.wtf/assets/yes/2.gif"
    })
  }))
  beforeEach(()=>{
    wrapper = shallowMount(Indecision);
    spyLog  = jest.spyOn(console, 'log');
    jest.clearAllMocks();
  });

  test("creando match con el snapshot", ()=>{
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Validad que un spy se ejecute una vez", async()=>{
    const spyAnswer = jest.spyOn(wrapper.vm, 'getAnswer');
    const input = wrapper.find("input");
    await input.setValue("Hola Mundo?");
    expect(spyLog).toHaveBeenCalledTimes(1);
    expect(spyAnswer).toHaveBeenCalled();
  })

  test("pruebas en getAnswer", async ()=>{
    await wrapper.vm.getAnswer();
    const img = wrapper.find('img');
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.image).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("Si!");
  })

  test("pruebas en getAnswer si falla", async()=>{
    fetch.mockImplementationOnce(()=>Promise.reject("API is Down"));
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No se pudo cargar Api");
  })
})