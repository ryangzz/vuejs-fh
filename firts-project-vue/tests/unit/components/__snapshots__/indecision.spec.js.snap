// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component creando match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" name="pregunta" id="pregunta" placeholder="Hazme una pregunta">
  <p>Recuerda terminar un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
